external/fv3fit/fv3fit/reservoir/autoencoder.py

```diff
@@ -26,20 +41,195 @@ def encode(self, x: Union[np.ndarray, tf.Tensor]) -> np.ndarray:
     def decode(self, latent_x: Union[np.ndarray, tf.Tensor]) -> np.ndarray:
         return self.decoder.predict(latent_x)
 
+
+@io.register("dense-autoencoder")
+class AutoencoderPredictor(Reloadable):
+    _ENCODER_NAME = "encoder.tf"
+    _DECODER_NAME = "decoder.tf"
+    _CONFIG_FILENAME = "autoencoder_config.yaml"
+
+    def __init__(
+        self,
+        input_variables: Iterable[Hashable],
+        output_variables: Iterable[Hashable],
+        model: Autoencoder,
+    ):
+        self.input_variables = input_variables
+        self.output_variables = output_variables
+        self.model = model
+
+    def predict(self, X: np.ndarray):
+        return self.model.predict(X)
+
+    def dump(self, path: str) -> None:
+        with put_dir(path) as path:
+            encoder_filename = os.path.join(path, self._ENCODER_NAME)
+            self.model.encoder.save(encoder_filename)
+            decoder_filename = os.path.join(path, self._DECODER_NAME)
+            self.model.decoder.save(decoder_filename)
+            with open(os.path.join(path, self._CONFIG_FILENAME), "w") as f:
+                f.write(
+                    yaml.dump(
+                        {
+                            "input_variables": self.input_variables,
+                            "output_variables": self.output_variables,
+                        }
+                    )
+                )
+
     @classmethod
-    def load(cls, path: str) -> "Autoencoder":
+    def load(cls, path: str) -> "AutoencoderPredictor":
         with get_dir(path) as model_path:
             encoder = tf.keras.models.load_model(
                 os.path.join(model_path, cls._ENCODER_NAME)
             )
             decoder = tf.keras.models.load_model(
                 os.path.join(model_path, cls._DECODER_NAME)
             )
-        return cls(encoder=encoder, decoder=decoder)
+            with open(os.path.join(path, cls._CONFIG_FILENAME), "r") as f:
+                config = yaml.load(f, Loader=yaml.Loader)
+        autoencoder = Autoencoder(encoder=encoder, decoder=decoder)
+        return cls(
+            input_variables=config["input_variables"],
+            output_variables=config["output_variables"],
+            model=autoencoder,
+        )
 
-    def dump(self, path: str) -> None:
-        with put_dir(path) as path:
-            encoder_filename = os.path.join(path, self._ENCODER_NAME)
-            self.encoder.save(encoder_filename)
-            decoder_filename = os.path.join(path, self._DECODER_NAME)
-            self.decoder.save(decoder_filename)
+
+@dataclasses.dataclass
+class DenseAutoencoderHyperparameters(Hyperparameters):
+    """
+    input_variables: variables to encode/decode
+    output_variables: variables to encode/decode, must match
+        input_variables
+    latent_dim_size: size of latent space that encoder transforms to
+    units: number of units in each dense layer
+    n_dense_layers: number of dense layers in *each* of encoder/decoder
```

Comment:
Suggestion (optional): Is it intentional that the hidden dimension size does not necessarily match the output latent dimension size? And that you don't need a way to e.g. ramp up or down the number of features (e.g. 50 inputs to 40 to 30 to 20 hidden features)? If you do need this extra configuration, you could use a `layer_units` attribute of type `List[str]` which gives the number of encoder units in order and number of decoder units in reverse order (and implicitly defines the number of layers).

Reply:
Yes, for the former I figured the capacitiy of the encoding/decoding hidden layers could be changed via either units or number of layer. Since our existing dense networks also are fixed to have the same units per layer I figured that was good enough for this use case as well.